src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/RegexIgnoreCaseTests.cs

```diff
@@ -0,0 +1,221 @@
+            Assert.True(regex.IsMatch(input));
+        }
+
+        public static IEnumerable<object[]> EnginesThatSupportBackreferences()
+        {
+            foreach(var engine in RegexHelpers.AvailableEngines)
+            {
+                if (engine == RegexEngine.NonBacktracking) // Nonbacktracking engine doesn't yet support backreferences.
+                    continue;
+                yield return new object[] { engine };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(EnginesThatSupportBackreferences))]
+        public async Task IgnoreCase_Behavior_Is_Constant(RegexEngine engine)
+        {
+            Regex regex;
+            using (_ = new ThreadCultureChange("en-US"))
+            {
+                regex = await RegexHelpers.GetRegexAsync(engine, @"(i)\1", RegexOptions.IgnoreCase);
+            }
+
+            using (_ = new ThreadCultureChange("tr-TR"))
+            {
+                // tr-TR culture doesn't consider 'I' and 'i' to be equal in ignore case, but en-US culture does.
+                // This test will validate that the backreference will use en-US culture even when current culture is
+                // set to tr-TR
+                Assert.True(regex.IsMatch("Ii"));
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(Characters_With_Common_Lowercase_Match_Backreference_Data))]
+        public async Task Characters_With_Common_Lowercase_Match_Backreference(RegexEngine engine, string pattern, string firstChar, string secondChar, string culture)
+        {
+            using var _ = new ThreadCultureChange(culture);
+            Regex regex = await RegexHelpers.GetRegexAsync(engine, pattern, RegexOptions.IgnoreCase);
+            Assert.True(regex.IsMatch($"{firstChar}{secondChar}"));
+            Assert.True(regex.IsMatch($"{secondChar}{firstChar}"));
+        }
+
+        [Theory]
+        [MemberData(nameof(EnginesThatSupportBackreferences))]
+        public async Task Ensure_CultureInvariant_Option_Is_Used_For_Backreferences(RegexEngine engine)
+        {
+            using var _ = new ThreadCultureChange("tr-TR");
+            Regex regex = await RegexHelpers.GetRegexAsync(engine, @"(.)\1", RegexOptions.CultureInvariant | RegexOptions.IgnoreCase);
+            // There is no mapping between 'i' and 'I' in tr-TR culture, so this test is validating that when passing CultureInvariant
+            // option, we will use InvariantCulture mappings for backreferences.
+            Assert.True(regex.IsMatch("iI"));
+        }
+
+        [Fact]
+        // This test creates a source generated engine for each of the ~870 cultures and ensures the result compiles. This test alone takes around 30
+        // seconds on a fast machine, so marking as OuterLoop.
+        [OuterLoop]
+        public async Task SourceGenerator_Supports_All_Cultures()
+        {
+            foreach (CultureInfo culture in CultureInfo.GetCultures(CultureTypes.AllCultures))
```

Comment:
would using a Parallel.For help here?

Reply:
Will try this but I don't expect it to help much. Running only both of these tests on my machine as it stands and with all 12 parallel threads take about 4 minutes. I will try Parallel.ForEach to see if that helps a bit.